// react-router-hook/lib/PropTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.routerHookContextShape = exports.renderPropsShape = exports.locationShape = exports.routerShape = exports.routesShape = exports.routeShape = exports.componentsShape = exports.componentShape = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var func = _propTypes2.default.func,
    object = _propTypes2.default.object,
    arrayOf = _propTypes2.default.arrayOf,
    oneOfType = _propTypes2.default.oneOfType,
    element = _propTypes2.default.element,
    shape = _propTypes2.default.shape,
    string = _propTypes2.default.string;
var componentShape = exports.componentShape = oneOfType([func, string]);
var componentsShape = exports.componentsShape = oneOfType([componentShape, object]);

var routeShape = exports.routeShape = oneOfType([object, element]);
var routesShape = exports.routesShape = oneOfType([routeShape, arrayOf(routeShape)]);

var routerShape = exports.routerShape = shape({
  push: func.isRequired,
  replace: func.isRequired,
  go: func.isRequired,
  goBack: func.isRequired,
  goForward: func.isRequired,
  setRouteLeaveHook: func.isRequired,
  isActive: func.isRequired
});

var locationShape = exports.locationShape = shape({
  pathname: string.isRequired,
  search: string.isRequired,
  state: object,
  action: string.isRequired,
  key: string
});

var renderPropsShape = exports.renderPropsShape = shape({
  location: locationShape,
  routes: routesShape,
  params: object,
  components: componentsShape,
  router: routerShape
});

var routerHookContextShape = exports.routerHookContextShape = shape({
  getComponentStatus: func,
  setComponentStatus: func,
  addLoadingListener: func
});
